feat(core): add support for i18n hydration

devknoll opened the pull request:
Add support for hydrating i18n blocks. This is accomplished by serializing information about selected ICU cases for a block during server-side rendering.

During hydration, this data is read and is used to traverse both an AST of the translated message and the DOM, in parallel, to map each LView with an RNode.

Finally, this mapping is used while nodes are being created (either via i18n or their respective instructions) to locate existing nodes.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

devknoll committed:
feat(core): add support for i18n hydration

Add support for hydrating i18n blocks. This is accomplished by serializing information about selected ICU cases for a block during server-side rendering.

During hydration, this data is read and is used to traverse both an AST of the translated message and the DOM, in parallel, to map each LView with an RNode.

Finally, this mapping is used while nodes are being created (either via i18n or their respective instructions) to locate existing nodes.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_hydration.ts:
```diff
@@ -0,0 +1,252 @@
+
+  const disconnectedNodes = hydrationInfo.disconnectedNodes ??=
+      new Set(hydrationInfo.data[DISCONNECTED_NODES] ?? []);
+  const i18nMap: Map<number, Node> = hydrationInfo.i18nNodes ??= new Map();
```
This might be sufficient to set the right type of the `i18nMap` const.
```suggestion
  const i18nMap = hydrationInfo.i18nNodes ??= new Map<number, Node>();
```

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_hydration.ts:
```diff
@@ -0,0 +1,252 @@
+  const i18nMap: Map<number, Node> = hydrationInfo.i18nNodes ??= new Map();
+  const caseQueue: number[] = hydrationInfo.data[I18N_DATA]?.[index - HEADER_OFFSET] ?? [];
+
+  interface HydrationState {
+    currentNode: Node|null;
+  }
```
Do we plan to have more fields in the state? If not, we can just define a `let` that points to the current node (or `null`)?

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_hydration.ts:
```diff
@@ -0,0 +1,252 @@
+              let childState = state;
+              let markOptions = MarkOptions.NONE;
+
+              if (containerSize == null) {
```
Do we always have `null` here or we may have `undefined`?
```suggestion
              if (containerSize === null) {
```

AndrewKushnir commented on packages/core/src/hydration/interfaces.ts:
```diff
@@ -92,6 +93,14 @@ export interface SerializedView {
   [DISCONNECTED_NODES]?: number[];
+
+  /**
+   * Serialized information about i18n views in a template.
```
```suggestion
   * Serialized information about i18n blocks in a template.
```

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -98,6 +98,37 @@ function locateOrCreateNode(
+  const native = locateI18nRNodeByIndex(hydrationInfo!, noOffsetIndex) as RNode;
+  if (ngDevMode) {
+    // TODO: Validate the element.
+    markRNodeAsClaimedByHydration(native);
```
Do we throw an error here is the `native` is not defined?

AndrewKushnir reviewed (commented):
@devknoll thanks for the PR, the change looks very good 👍 I did a first review pass and left some comments. Please let me know once the feedback is addressed, I will do another pass. Thank you.

pschuler21 commented:
Very much looking forward to this being merged

![OPcxBQchmR2EYK7abQ-ezgif com-optimize](https://github.com/angular/angular/assets/94828331/c6a66b51-17d6-4d7d-b4dc-4f0bde869285)

devknoll commented on packages/core/src/hydration/api.ts:
```diff
@@ -62,6 +64,9 @@ function enableHydrationRuntimeSupport() {
     enableFindMatchingDehydratedViewImpl();
     enableApplyRootElementTransformImpl();
+    enableLocateOrCreateI18nNodeImpl();
+    enableTrySerializeI18nBlockImpl();
```
Ahh, whoops, good catch.

devknoll commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -299,6 +300,51 @@ function appendDisconnectedNodeIndex(ngh: SerializedView, tNode: TNode) {
+ * that seperate logical text blocks in the DOM remain separate after
+ * serialization.
+ */
+function processTextNodeBeforeSerialization(context: HydrationContext, node: RNode) {
```
I gave this a try, but the diff didn't seem particularly useful. If you feel strongly about it, let me know and I'll include the commit here.

devknoll commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,334 @@
+        const selectedCase = state.isConnected ? context.caseQueue.shift()! : null;
+        const childState = {currentNode: null, isConnected: false};
+
+        // We traverse through each case, even if it's not selected,
```
```suggestion
        // We traverse through each case, even if it's not active,
```

devknoll commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,334 @@
+        break;
+
+      case I18nNodeKind.ICU:
+        // If the current node is attached, we need to pop the next case from the
+        // queue, so that the active case is also considered attached.
```
```suggestion
        // If the current node is connected, we need to pop the next case from the
        // queue, so that the active case is also considered connected.
```

devknoll commented:
@AndrewKushnir Thanks for your detailed feedback. All of the feedback should be addressed (save for the nit that I left a comment on), so this should be ready for another review when you get a chance.

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -159,6 +161,9 @@ function annotateLContainerForHydration(lContainer: LContainer, context: Hydrati
  */
 export function annotateForHydration(appRef: ApplicationRef, doc: Document) {
+  const injector = appRef.injector;
+  const enableI18nHydration = injector.get(IS_I18N_HYDRATION_ENABLED, false);
```
nit: I'd propose to name this const similar to the config token:
```suggestion
  const isI18nHydrationEnabled = injector.get(IS_I18N_HYDRATION_ENABLED, false);
```

AndrewKushnir commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,332 @@
+ *
+ * @param lView lView with the i18n block
+ * @param index index of the i18n block in the lView
+ * @param processTextNode callback run for text nodes
```
The docs for this argument are outdated (we've switched to passing a context).

AndrewKushnir reviewed (commented):
@devknoll thanks for addressing the feedback! 👍 

I've left a few more comments and will do another pass on the main algorithm tomorrow.

AndrewKushnir commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,325 @@
+    const rootNode = locateNextRNode(hydrationInfo, tView, lView, tNode) as Node | null;
+
+    const disconnectedNodes = initDisconnectedNodes(hydrationInfo) ?? new Set();
+    const i18nMap = hydrationInfo.i18nNodes ??= new Map<number, RNode|null>();
```
I'd like to propose that we replace `i18nMap` with `i18nNodes`, so it's easier to make a "connection" to the `i18nNodes` field in `hydrationInfo` object.
```suggestion
    const i18nNodes = hydrationInfo.i18nNodes ??= new Map<number, RNode|null>();
```

devknoll replied:
Ah, good catch, my mistake.

AndrewKushnir commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,325 @@
+    switch (nodeOrNodes.kind) {
+      case I18nNodeKind.TEXT:
+        // Claim a text node for hydration
+        markHydrationRoot(context, state, nodeOrNodes, MarkOptions.CLAIM);
```
It's a bit hard to reason about the state (how we advance it), since the state modifications are "hidden" inside of the `markHydrationRoot` and `skipHydrationNodes` functions. I'd like to propose that we make state modification more explicit, i.e. do something like this:
```suggestion
        const currentNode = markHydrationRoot(context, state, nodeOrNodes);
        advanceCurrentNode(state, currentNode?.nextSibling);
```
Similarly, for the `skipHydrationNodes`:
```
        const nextNode = skipHydrationNodes(state, containerSize! + 1);
        advanceCurrentNode(state, nextNode);
```
In this case we can also drop `MarkOptions`.

AndrewKushnir replied:
With this change, we can also update the names of the `prepareForHydration` (quite generic) and `markHydrationRoot` functions, for example:

- `prepareForHydration` -> `collectI18nNodesFromDom`
- `markHydrationRoot` -> `appendI18nNodeToCollection`

devknoll replied:
In this case, the `advanceCurrentNode` isn't actually doing any advancing (since we're just passing in the node), so should it just be `setCurrentNode`?

AndrewKushnir replied:
Yeah, `setCurrentNode` sounds great 👍

AndrewKushnir commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,325 @@
+ * sometimes we might encounter null nodes. In those cases, we want to
+ * defer to downstream error handling to provide proper context.
+ */
+function skipHydrationNodes(state: I18nHydrationState, skip: number) {
```
```suggestion
function skipSiblingNodes(state: I18nHydrationState, skip: number) {
```

AndrewKushnir reviewed (commented):
@devknoll Thanks for the update, I did another pass over the main logic and proposed a few changes.

devknoll commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,325 @@
+
+  const firstAstNode = tI18n.ast[0];
+  if (firstAstNode) {
+    // We begin hydration at the RNode for the first child. Top-level AST nodes will always
+    // have a valid TNode, so we can use the normal `locateNextRNode` to find it. It's OK if
```
Yes, all of the top-level nodes of the AST will have a valid TNode.

`i18nStartFirstCreatePass` created the top-level AST. There are only three potential cases:

1. Text, in which case a text TNode is created by `i18nStartFirstCreatePassProcessTextNode`
2. ICU, in which case an ICU TNode is created directly (via `createTNodeAndAddOpCode`)
3. Placeholder, in which case the TNode is created directly by `createTNodePlaceholder`

More concisely, the only times we _don't_ have a TNode are for the content of ICUs, where we just allocated directly without creating a corresponding TNode.

I'll update the comment with this more concise explanation.

devknoll commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -0,0 +1,325 @@
+        {hydrationInfo, i18nMap, disconnectedNodes, caseQueue},
+        {currentNode: rootNode, isConnected: true}, tI18n.ast);
+
+    // During i18n hydration, we may have discovered disconnected nodes that weren't
+    // serialized, so we need to write the expected value back.
+    hydrationInfo.disconnectedNodes = disconnectedNodes.size === 0 ? null : disconnectedNodes;
```
Today, we **only** serialize disconnected nodes if they are also a projected node.

That's fine, because in the case of i18n, this information would be redundant, as it can be derived via the active ICU cases.

And that's essentially what this comment is trying to say: disconnected nodes from inactive ICU cases wouldn't have be serialized, but we still want them in the disconnected set (because they are disconnected from the DOM).

I'll clarify the comment.

devknoll committed:
fix(core): support content projection and VCRs in i18n

Modifies the i18n pre-hydration logic to correctly support content projection and elements that act as view containers.

devknoll commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -40,9 +40,11 @@ export function isI18nHydrationSupportEnabled() {
  *
  * @param lView lView with the i18n block
  * @param index index of the i18n block in the lView
+ * @param parentTNode TNode of the parent of the i18n block
```
Missing `subTemplateIndex` here

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -318,6 +319,7 @@ function appendDisconnectedNodeIndex(ngh: SerializedView, tNode: TNode) {
 function serializeLView(lView: LView, context: HydrationContext): SerializedView {
   const ngh: SerializedView = {};
   const tView = lView[TVIEW];
+  const i18nChildren = collectI18nChildren(tView);
```
I think we can do this calculation for a TView once and store it in the `context` (as a `Map<TView, Set<number>|null>`), since the data would not change for different instances. That'd help to optimize cases when we stamp out component instances in a loop.

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -473,11 +475,12 @@ function serializeLView(lView: LView, context: HydrationContext): SerializedView
  *     connection to identify the location of a node.
  */
-function conditionallyAnnotateNodePath(ngh: SerializedView, tNode: TNode, lView: LView<unknown>) {
+function conditionallyAnnotateNodePath(
```
It'd be great to add a note on why we pass `i18nChildren` here. I'm also thinking if we should rename the argument in this function (and other affected fns that it invokes) to something like `excludedParentNodes`, which conveys the message a bit better.

AndrewKushnir commented on packages/core/src/hydration/node_lookup_utils.ts:
```diff
@@ -57,12 +57,28 @@ export function locateI18nRNodeByIndex<T extends RNode>(
+ * @param hydrationInfo The hydration annotation data
+ * @param lView the current lView
+ * @param noOffsetIndex the instruction index
+ * @returns an RNode that corresponds to the instruction index
```
We can add that `null` can also be returned: "... or null in case no RNode was found".

AndrewKushnir commented on packages/core/src/hydration/node_lookup_utils.ts:
```diff
@@ -57,12 +57,28 @@ export function locateI18nRNodeByIndex<T extends RNode>(
+ * @param noOffsetIndex the instruction index
+ * @returns an RNode that corresponds to the instruction index
+ */
+export function tryLocateRNodeByPath<T extends RNode>(
```
We don't seem to use generic anywhere in arguments and it'd always be inferred as `RNode`, so we can just drop the generic.

AndrewKushnir reviewed (commented):
@devknoll thanks for additional changes. I've looked at the content projection logic and left some comments. Will take a look at the ViewContainer part next and share more feedback.

devknoll committed:
fix(core): prevent i18n hydration from cleaning projected nodes

Rework the i18n cleanup behavior to more closely match that of containers. Specifically, we assume that nodes are going to be claimed unless they are part of a branching ICU block.

During hydration, we then track which ICU case was active at serialization time, and which was active during hydration. Any remaining cases that weren't used during hydration are then cleaned up.

AndrewKushnir commented on packages/core/src/hydration/i18n.ts:
```diff
@@ -206,40 +269,54 @@ function prepareI18nBlockForHydrationImpl(lView: LView, index: number) {
-    // new set, we need to be sure to write the expected value back to the DehydratedView.
-    hydrationInfo.disconnectedNodes = disconnectedNodes.size === 0 ? null : disconnectedNodes;
+  function findHydrationRoot() {
+    if (subTemplateIndex < 0) {
```
We can use the `isRootTemplateMessage` helper here (which would also help with readability).

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -103,9 +106,23 @@ let _locateOrCreateNode: typeof locateOrCreateNodeImpl = (lView, index, textOrNa
+
+  const native = locateI18nRNodeByIndex(hydrationInfo!, noOffsetIndex) as RNode;
+
+  // TODO: Improve error handling
```
nit: it'd be great to provide a bit more info on what we need to improve here, so it's easier to pick it up later.

AndrewKushnir reviewed (approved):
@devknoll looks great 👍

devknoll commented:
[Presubmit](http://test/OCL:619997839:BASE:619997885:1711649651073:36302ac3)

atscott commented:
This PR was merged into the repository by commit a5fa279b6e9f5ab4005d6d33107f0e1bb48d05de.
